rasa/shared/utils/common.py

```diff
@@ -66,6 +68,36 @@ def _lazyprop(self):
     return _lazyprop
 
 
+def cached_method(f: Callable[..., Any]) -> Callable[..., Any]:
+    """Caches the first method call result.
+
+    Works for `async` and `sync` methods. Don't apply this to functions.
+
+    Returns:
+        The return value which the method gave for the first call.
+    """
+    if asyncio.iscoroutinefunction(f):
+
+        @functools.wraps(f)
+        async def decorated(self: object, *args: Any, **kwargs: Any) -> Any:
+            cache_name = f"_cached_{self.__class__.__name__}_{f.__name__}"
```

Comment:
don't want to play devils advocate, but doesn't this reuse the same cache result for different argument values? (e.g. language parameter when loading nlu)

Reply:
maybe we should use a std lib implementation for this? `functools.lru_cache`

Reply:
`lru_cache` doesn't work with async. 

> but doesn't this reuse the same cache result for different argument values?

Yes, but I think it's fine. A new importer should be created if used with different parameters (we never user it with the same parameters in one go)

Reply:
mhm it is a source of very hidden bugs though - should we move the arguments to init then?

Reply:
yes very true - if that is possible I think that would be perfect

Reply:
So it ended up being better, but not perfect (see my other comment). At this point we're solving two problems:
* caching the calls
* having an async cache function

I still think the decorator and its currently relatively lightweight implementation are the best solution. The next best thing would be to move all arguments to `__init__` which is both implementation pain + pain for users (potentially even Rasa X, don't know how the training data reading is implemented there for 2.0)